src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/SymbolKey/SymbolKey.SymbolKeyReader.cs

```diff
@@ -191,6 +191,16 @@ public PooledArrayBuilder<RefKind> ReadRefKindArray()
                 => ReadArray(_readRefKind, out _);
 
             public PooledArrayBuilder<T> ReadArray<T>(ReadFunction<T> readFunction, out string? failureReason)
+                => ReadArray(readFunction, beforeRead: static (_, _) => false, afterRead: static _ => { }, data: false, out failureReason);
+
+            public PooledArrayBuilder<T> ReadArray<T, TData1, TData2>(
+                ReadFunction<T> readFunction,
+                Func<TData1, int, TData2> beforeRead,
+                Action<TData2> afterRead,
+                TData1 data,
+                out string? failureReason)
+                where TData1 : struct
+                where TData2 : struct
```

Comment:
pretty ugly.  but it allows us to read each element, setting/clearing contextual information for each element in a non-allocating fashion.

ALternatives woudl be to hand write this loop for reading symbols (thus not sharing the loop logic when reading arrays of anything else).  We could also potentially simplify the two functions by not having them send data between each other.  however, the value of sending teh data is that we can do some very useful asserst in hte 'afterRead' call that i would like to preserve.